### PR TITLE
feat: Support non-browser requests for development/staging environments

### PR DESCRIPTION
**Pull request by chanioxaris**
## Type of change

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 🌟 New feature (non-breaking change which adds functionality)
- [ ] 🔨 Breaking change (fix or feature that would cause existing functionality)
- [ ] 📖 Docs change / refactoring / dependency upgrade to change)

## Description

This PR contains the below changes:
1. For development/staging environments, try to detect if that's a non-browser request (e.g. script) based on the request `User Agent` and prevent interstitial rendering
2. Take into account the `Authorization` header every time it exists in the request

### Related Issue (optional)

### CODE REVIEW

File: src/Clerk.ts [hunk 1]
```diff
@@ -330,21 +330,16 @@ export default class Clerk {
...
 
         // HEADER AUTHENTICATION
-        if (headerToken && !decodeToken(this, headerToken)) {
-          return signedOut();
-        }
-
-        if (!headerToken && isCrossOriginRequest(req)) {
-          return signedOut();
-        }
+        if (headerToken !== undefined) {
```
Comment (gkats) [inline src/Clerk.ts:338]: :upside_down_face: You can probably return early here, unless I got confused.

```js
// Assuming that you want to explicitly check for undefined and not null or ''
if (headerToken === undefined) return signedOut()
if (!decodeToken) return signedOut()
// do the work to verify token and create session
```

File: src/Clerk.ts [hunk 2]
```diff
@@ -330,21 +330,16 @@ export default class Clerk {
         Logger.debug(`cookieToken: ${cookieToken}`);
         const clientUat = cookies.get('__client_uat');
         Logger.debug(`clientUat: ${clientUat}`);
-        let headerToken = (req.headers['Authorization'] ||
-          req.headers['authorization']) as string;
+        let headerToken = req.headers.authorization;
```
Comment (gkats) [inline src/Clerk.ts:333]: :question: Unrelated to this PR, this is vanilla Node.js right? Why are we using the `Request` type from expressjs?

File: src/Clerk.ts [hunk 3]
```diff
@@ -360,6 +355,19 @@ export default class Clerk {
           return next();
         }
 
+        // In development or staging environments only, based on the request's
+        // User Agent, detect non-browser requests (e.g. scripts). If there
+        // is no Authorization header, consider the user as signed out and
+        // prevent interstitial rendering
+        if (isDevelopmentOrStaging(this._restClient.apiKey) && !req.headers['user-agent']?.startsWith('Mozilla/')) {
```
Comment (gkats) [inline src/Clerk.ts:362]: I'm worried that checking for "Mozilla" user agent might not be enough to guarantee this is a browser request. Why is this necessary?

File: src/Clerk.ts (hunk 1 above)
Comment (chanioxaris) [reply src/Clerk.ts:338]: The purpose of this change is to take into account the `Authorization` header if exists in the incoming request.

If it doesn't (undefined) then we should move in the next steps below and probably the cookie authentication. We can't return early if the `Authorization` header doesn't exist, as the `__session` cookie might exist

File: src/Clerk.ts (hunk 3 above)
Comment (SokratisVidros) [reply src/Clerk.ts:362]: For browsers, almost 99.9% of the user-agent strings start with Mozilla. There are some exceptions, such as the old Opera version or the web browser of Apple TV v4. All in all, this is a cheap best-effort check to prevent rendering the interstitial if the HTTP Authorization header is empty. 

For example, if an app API is called from a CLI script, rendering the interstitial is confusing. With this fix, we will return a 401 instead.

Does that make sense?

File: src/Clerk.ts (hunk 2 above)
Comment (SokratisVidros) [reply src/Clerk.ts:333]: @chanioxaris Have we tested that this is case insensitive in popular node frameworks such as Express, Next or Fastify?

Comment (gkats) [reply src/Clerk.ts:333]: The `req` object should make it case-insensitive. Express offers the `req.get('authorization')` and `req.header('authorization')` which would theoretically work. I'm guessing the same happens in Node.js? A simple test with 

```
curl -H 'Authorization: Foo'
curl -H 'authorization: Foo'
```
will do.

File: src/Clerk.ts (hunk 3 above)
Comment (gkats) [reply src/Clerk.ts:362]: It does, thanks for explaining. There's certainly no guarantee. I guess the rationale is "it's fine since it's only for dev".

File: src/Clerk.ts (hunk 2 above)
Comment (chanioxaris) [reply src/Clerk.ts:333]: I have test it with a Next.js app for the moment, i will also need to have a look into Express and Fastify.

Comment (chanioxaris) [reply src/Clerk.ts:333]: @SokratisVidros Tested with Next.js (`v11.0.1`), Express (`4.17.1`) and Fastify (`v3.25.0`) and working as expected

**Review by SokratisVidros — APPROVED** [review]

**Comment by SokratisVidros** [comment]
@igneel64 Are you OK to merge this?
